plugins/virtual_folders/server/__init__.py

```diff
@@ -0,0 +1,125 @@
+import json
+from bson import json_util
+from girder import events
+from girder.api import access, rest
+from girder.api.v1.folder import Folder as FolderResource
+from girder.constants import AccessType
+from girder.exceptions import ValidationException
+from girder.models.folder import Folder
+from girder.models.item import Item
+
+
+def _validateFolder(event):
+    doc = event.info
+
+    if 'isVirtual' in doc and not isinstance(doc['isVirtual'], bool):
+        raise ValidationException('The isVirtual field must be boolean.', field='isVirtual')
+
+    if doc.get('isVirtual'):
+        # Make sure it doesn't have children
+        if list(Folder().childItems(doc, limit=1)):
+            raise ValidationException(
+                'Virtual folders may not contain child items.', field='isVirtual')
+        if list(Folder().find({
+            'parentId': doc['_id'],
+            'parentCollection': 'folder'
+        }, limit=1)):
+            raise ValidationException(
+                'Virtual folders may not contain child folders.', field='isVirtual')
+    if doc['parentCollection'] == 'folder':
+        parent = Folder().load(event.info['parentId'], force=True, exc=True)
+        if parent.get('isVirtual'):
+            raise ValidationException(
+                'You may not place folders under a virtual folder.', field='folderId')
+
+    if 'virtualItemsQuery' in doc:
+        try:
+            json.loads(doc['virtualItemsQuery'])
+        except (TypeError, ValueError):
+            raise ValidationException(
+                'The virtual items query must be valid JSON.', field='virtualItemsQuery')
+
+    if 'virtualItemsSort' in doc:
+        try:
+            json.loads(doc['virtualItemsSort'])
```

Comment:
Should we go further and ensure that this could be used in a mongo query?  Or, just give an example in the param description (I never remember that it is of the form `[[(field), (direction)], [(field), (direction)]]`.

Reply:
I think that'd be a good feature, although I might want to do it on a different PR.